File: inscription-service/src/main/java/com/services/InscriptionService.java
package com.services;

import java.util.List;

import com.dto.CreateInscriptionRequest;
import com.dto.InscriptionDTO;
import com.dto.InscriptionStatusDTO;
import com.dto.ReinscriptionRequest;
import com.dto.UpdateInscriptionRequest;
import com.dto.ValidateInscriptionRequest;
import com.enums.InscriptionStatus;

public interface InscriptionService {

    List<InscriptionDTO> getAllInscriptions();
    InscriptionDTO getInscriptionById(String id);
    InscriptionDTO createInscription(CreateInscriptionRequest request);
    InscriptionDTO updateInscription(String id, UpdateInscriptionRequest request);
    void deleteInscription(String id);
    InscriptionDTO validateByDirecteur(String id, ValidateInscriptionRequest request);
    InscriptionDTO validateByAdmin(String id, ValidateInscriptionRequest request);
    InscriptionStatusDTO getInscriptionStatus(String id);
    List<InscriptionDTO> getInscriptionsByDoctorant(String doctorantId);
    List<InscriptionDTO> getInscriptionsByStatus(InscriptionStatus status);
    InscriptionDTO createReinscription(ReinscriptionRequest request);
}

File: doctorat-frontend/src/app/services/inscription.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiConfigService } from './api-config.service';
import {
  Inscription,
  CreateInscriptionRequest,
  UpdateInscriptionRequest,
  ValidateInscriptionRequest,
  ReinscriptionRequest,
  InscriptionResponse,
  InscriptionListResponse,
  InscriptionStatus,
  InscriptionStatusDTO
} from '../models/inscription.model';

@Injectable({
  providedIn: 'root'
})
export class InscriptionService {
  constructor(
    private http: HttpClient,
    private apiConfig: ApiConfigService
  ) { }

  getAllInscriptions(): Observable<InscriptionListResponse> {
    return this.http.get<InscriptionListResponse>(this.apiConfig.inscriptionsUrl);
  }

  getInscriptionById(id: string): Observable<InscriptionResponse> {
    return this.http.get<InscriptionResponse>(`${this.apiConfig.inscriptionsUrl}/${id}`);
  }

  createInscription(request: CreateInscriptionRequest): Observable<InscriptionResponse> {
    return this.http.post<InscriptionResponse>(this.apiConfig.inscriptionsUrl, request);
  }

  updateInscription(id: string, request: UpdateInscriptionRequest): Observable<InscriptionResponse> {
    return this.http.put<InscriptionResponse>(`${this.apiConfig.inscriptionsUrl}/${id}`, request);
  }

  deleteInscription(id: string): Observable<any> {
    return this.http.delete(`${this.apiConfig.inscriptionsUrl}/${id}`);
  }

  validateByDirecteur(id: string, request: ValidateInscriptionRequest): Observable<InscriptionResponse> {
    return this.http.put<InscriptionResponse>(
      `${this.apiConfig.inscriptionsUrl}/${id}/validate/directeur`,
      request
    );
  }

  validateByAdmin(id: string, request: ValidateInscriptionRequest): Observable<InscriptionResponse> {
    return this.http.put<InscriptionResponse>(
      `${this.apiConfig.inscriptionsUrl}/${id}/validate/admin`,
      request
    );
  }

  getInscriptionStatus(id: string): Observable<InscriptionStatusDTO> {
    return this.http.get<InscriptionStatusDTO>(`${this.apiConfig.inscriptionsUrl}/${id}/status`);
  }

  getInscriptionsByDoctorant(doctorantId: string): Observable<InscriptionListResponse> {
    return this.http.get<InscriptionListResponse>(
      `${this.apiConfig.inscriptionsUrl}/doctorant/${doctorantId}`
    );
  }

  getInscriptionsByStatus(status: InscriptionStatus): Observable<InscriptionListResponse> {
    return this.http.get<InscriptionListResponse>(
      `${this.apiConfig.inscriptionsUrl}/status/${status}`
    );
  }

  createReinscription(request: ReinscriptionRequest): Observable<InscriptionResponse> {
    return this.http.post<InscriptionResponse>(
      `${this.apiConfig.inscriptionsUrl}/reinscriptions`,
      request
    );
  }

  healthCheck(): Observable<any> {
    return this.http.get(`${this.apiConfig.inscriptionsUrl}/health`);
  }
}

